plugins/extraction/json.ml

```diff
@@ -0,0 +1,274 @@
+        ("name", json_int i)
+      ])
+  | Tglob (r, l) -> json_dict [
+      ("what", json_str "type:glob");
+      ("name", json_global Type r);
+      ("args", json_list (List.map (json_type vl) l))
+    ]
+  | Tarr (t1,t2) -> json_dict [
+      ("what", json_str "type:arrow");
+      ("left", json_type vl t1);
+      ("right", json_type vl t2)
+    ]
+  | Tdummy _ -> json_dict [("what", json_str "type:dummy")]
+  | Tunknown -> json_dict [("what", json_str "type:unknown")]
+  | Taxiom -> json_dict [("what", json_str "type:axiom")]
+
+
+(*s Pretty-printing of expressions. *)
+
+let rec json_expr env = function
+  | MLrel n -> json_dict [
+      ("what", json_str "expr:rel");
+      ("name", json_id (get_db_name n env))
+    ]
+  | MLapp (f, args) -> json_dict [
+      ("what", json_str "expr:apply");
+      ("func", json_expr env f);
+      ("args", json_list (List.map (json_expr env) args))
+    ]
+  | MLlam _ as a ->
+    let fl, a' = collect_lams a in
+    let fl, env' = push_vars (List.map id_of_mlid fl) env in
+    json_dict [
+      ("what", json_str "expr:lambda");
+      ("argnames", json_list (List.map json_id (List.rev fl)));
+      ("body", json_expr env' a')
+    ]
+  | MLletin (id, a1, a2) ->
+    let i, env' = push_vars [id_of_mlid id] env in
+    json_dict [
+      ("what", json_str "expr:let");
+      ("name", json_id (List.hd i));
+      ("nameval", json_expr env a1);
+      ("body", json_expr env' a2)
+    ]
+  | MLglob r -> json_dict [
+      ("what", json_str "expr:global");
+      ("name", json_global Term r)
+    ]
+  | MLcons (_, r, a) -> json_dict [
+      ("what", json_str "expr:constructor");
+      ("name", json_global Cons r);
+      ("args", json_list (List.map (json_expr env) a))
+    ]
+  | MLtuple l -> json_dict [
+      ("what", json_str "expr:tuple");
+      ("items", json_list (List.map (json_expr env) l))
+    ]
+  | MLcase (typ, t, pv) -> json_dict [
+      ("what", json_str "expr:case");
+      ("expr", json_expr env t);
+      ("cases", json_listarr (Array.map (fun x -> json_one_pat env x) pv))
+    ]
+  | MLfix (i, ids, defs) ->
+    let ids', env' = push_vars (List.rev (Array.to_list ids)) env in
+    let ids' = Array.of_list (List.rev ids') in
+    json_dict [
+      ("what", json_str "expr:fix");
+      ("funcs", json_listarr (Array.map (fun (fi, ti) ->
+        json_dict [
+          ("what", json_str "fix:item");
+          ("name", json_id fi);
+          ("body", json_function env' ti)
+        ]) (Array.map2 (fun a b -> a,b) ids' defs)))
+    ]
+  | MLexn s -> json_dict [
+      ("what", json_str "expr:exception");
+      ("msg", json_str s)
+    ]
+  | MLdummy -> json_dict [("what", json_str "expr:dummy")]
+  | MLmagic a -> json_dict [
+      ("what", json_str "expr:coerce");
+      ("value", json_expr env a)
+    ]
+  | MLaxiom -> json_dict [("what", json_str "expr:axiom")]
+
+and json_one_pat env (ids,p,t) =
+  let ids', env' = push_vars (List.rev_map id_of_mlid ids) env in json_dict [
```

Comment:
I think that's indeed largely due to how the MiniML data structures already look like.  I think I copied those List.rev_map, push_vars, etc snippets from the existing extraction code for Haskell/Scheme/OCaml (e.g., see haskell.ml, scheme.ml, and ocaml.ml under plugins/extraction).